src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPDocumentSynchronizer.cs

```diff
@@ -2,88 +2,129 @@
+                {
+                    // Currently tracked synchronizing context is not sufficient, need to update a new one.
+                    documentContext.UpdateSynchronizingContext(requiredHostDocumentVersion, cancellationToken);
+                }
+                else
+                {
+                    // Already have a synchronizing context for this type of request, memoize the results.
+                }
+
+                var synchronizingContext = documentContext.SynchronizingContext;
+                return synchronizingContext.OnSynchronizedAsync;
             }
+        }
 
-            if (document.Version == virtualDocument.HostDocumentSyncVersion)
+        private void VirtualDocumentBuffer_PostChanged(object sender, EventArgs e)
+        {
+            var textBuffer = (ITextBuffer)sender;
+            if (!_fileUriProvider.TryGet(textBuffer, out var virtualDocumentUri))
             {
-                // Already synchronized
-                return true;
+                return;
             }
 
-            var synchronizingContext = _synchronizingContexts.AddOrUpdate(
-                virtualDocument.Uri,
-                (uri) => new DocumentSynchronizingContext(virtualDocument, document.Version, _synchronizationTimeout, cancellationToken),
-                (uri, existingContext) =>
+            lock (DocumentContextLock)
+            {
+                if (!_virtualDocumentContexts.TryGetValue(virtualDocumentUri, out var documentContext))
                 {
-                    if (virtualDocument == existingContext.VirtualDocument &&
-                        document.Version == existingContext.ExpectedHostDocumentVersion)
-                    {
-                        // Already contain a synchronizing context that represents this request and it's in-process of being calculated.
-                        return existingContext;
-                    }
+                    return;
+                }
 
-                    // Cancel old request
-                    existingContext.SetSynchronized(false);
-                    return new DocumentSynchronizingContext(virtualDocument, document.Version, _synchronizationTimeout, cancellationToken);
-                });
+                if (!textBuffer.TryGetHostDocumentSyncVersion(out var hostDocumentVersion))
+                {
+                    return;
+                }
 
-            var result = await _joinableTaskFactory.RunAsync(() => synchronizingContext.OnSynchronizedAsync);
+                documentContext.UpdateSeenDocumentVersion(hostDocumentVersion);
 
-            _synchronizingContexts.TryRemove(virtualDocument.Uri, out _);
+                var synchronizingContext = documentContext.SynchronizingContext;
+                if (synchronizingContext == null)
+                {
+                    // No active synchronizing context for this document.
```

Comment:
What scenarios might lead to this?

Reply:
This is the default flow of this API. Aka, you have a document open and you're typing in it. This will get repeatedly hit because:

1. We track every document that our `LSPDocumentManager` knows about. By track I mean we attach to their buffers `PostChanged` event
2. Synchronizing operations only occur when an LSP event occurs such as a completion, hover etc.

Sooo because we track every document rando typing will result in us repeatedly looking for a synchronizing context which usually won't be there.